src/linquantarrays.jl

```diff
@@ -10,84 +13,164 @@ Base.size(QA::LinQuantArray) = size(QA.A)
 
     Q = similar(A,T)                        # preallocate
 
-    # map minimum to 0x0, maximum to 0xff...ff
+    # map minimum to typemin(T), maximum to typemax(t)
     @inbounds for i in eachindex(Q)
-        Q[i] = round((A[i]-Amin)*Δ)
+        Q[i] = round((A[i]-Amin)*Δ⁻¹ + Tmin)
     end
 
     return LinQuantArray{T,ndims(Q)}(Q,Amin,Amax)
 end
 
-# define 8, 16, 24 and 32 bit
+function LinQuantization(
+    ::Type{T},
+    A::AbstractArray,
+    extrema::Tuple = extrema(A),
+) where {T<:Integer}
+    all(isfinite.(A)) || throw(DomainError("Linear quantization only in (-∞,∞)"))
+
+    # minimum-maximum range of values
+    Amin, Amax = Float64.(extrema)    
+    
+    # minimum and maximum representable value of type T
+    Tmin, Tmax = Float64(typemin(T)), Float64(typemax(T))
+
+    # inverse spacing, set to zero for no range
+    Δ⁻¹ = Amin == Amax ? zero(Float64) : (Tmax-Tmin)/(Amax-Amin)
+    
+    # preallocate
+    Q = similar(A, T)                        
+
+    # map minimum to typemin(T), maximum to typemax(t)
+    # clamp to [Amin,Amax] removing out-of-range values
+    @inbounds for i in eachindex(Q)
+        Q[i] = round(T, clamp((A[i]-Amin)*Δ⁻¹ + Tmin, Tmin, Tmax))
+    end
+
+    return LinQuantArray{T,ndims(Q)}(Q,Amin,Amax)
+end
+
+# define for unsigned integers of  8, 16, 24 and 32 bit 
 LinQuant8Array(A::AbstractArray{T,N}) where {T,N} = LinQuantization(UInt8,A)
 LinQuant16Array(A::AbstractArray{T,N}) where {T,N} = LinQuantization(UInt16,A)
 LinQuant24Array(A::AbstractArray{T,N}) where {T,N} = LinQuantization(UInt24,A)
 LinQuant32Array(A::AbstractArray{T,N}) where {T,N} = LinQuantization(UInt32,A)
 
+# define for unsigned integers of  8, 16, 24 and 32 bit 
+LinQuantUInt8Array(A::AbstractArray{T,N}, e::Option{Tuple}) where {T,N} = LinQuantization(UInt8,A,e)
+LinQuantUInt16Array(A::AbstractArray{T,N}, e::Option{Tuple}) where {T,N} = LinQuantization(UInt16,A,e)
+LinQuantUInt24Array(A::AbstractArray{T,N}, e::Option{Tuple}) where {T,N} = LinQuantization(UInt24,A,e)
+LinQuantUInt32Array(A::AbstractArray{T,N}, e::Option{Tuple}) where {T,N} = LinQuantization(UInt32,A,e)
+
+# define for signed integers of  8, 16, 24 and 32 bit
+LinQuantInt8Array(A::AbstractArray{T,N}, e::Option{Tuple}) where {T,N} = LinQuantization(Int8,A,e)
+LinQuantInt16Array(A::AbstractArray{T,N}, e::Option{Tuple}) where {T,N} = LinQuantization(Int16,A,e)
+LinQuantInt24Array(A::AbstractArray{T,N}, e::Option{Tuple}) where {T,N} = LinQuantization(Int24,A,e)
+LinQuantInt32Array(A::AbstractArray{T,N}, e::Option{Tuple}) where {T,N} = LinQuantization(Int32,A,e)
+
+
 """De-quantise a LinQuantArray into floats."""
-function Base.Array{T}(n::Integer,Q::LinQuantArray) where {T<:AbstractFloat}
-    Qmin = Q.min                # min as Float64
-    Qmax = Q.max                # max as Float64
-    Δ = (Qmax-Qmin)/(2^n-1)     # linear spacing
+function Base.Array{U}(n::Integer, Q::LinQuantArray) where {U<:AbstractFloat}
```

Comment:
yeah I think it makes sense to remove the `n` because that information is in the `eltype` of `Q`.